node/rpc.go

```diff
@@ -0,0 +1,150 @@
+package node
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/rs/cors"
+	cfg "github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/libs/log"
+	tmpubsub "github.com/tendermint/tendermint/libs/pubsub"
+	"github.com/tendermint/tendermint/libs/strings"
+	rpccore "github.com/tendermint/tendermint/rpc/core"
+	grpccore "github.com/tendermint/tendermint/rpc/grpc"
+	rpcserver "github.com/tendermint/tendermint/rpc/jsonrpc/server"
+	"github.com/tendermint/tendermint/types"
+)
+
+func startGRPCServer(rpcConfig *cfg.RPCConfig,
+	env *rpccore.Environment,
+	logger log.Logger) (net.Listener, error) {
+	// we expose a simplified api over grpc for convenience to app devs
+	listener, err := rpcserver.Listen(rpcConfig.GRPCListenAddress, rpcConfig.GRPCMaxOpenConnections)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		if err := grpccore.StartGRPCServer(env, listener); err != nil {
+			logger.Error("Error starting gRPC server", "err", err)
+		}
+	}()
+	return listener, nil
+}
+
+func startHTTPRPCServer(rpcConfig *cfg.RPCConfig,
+	logger log.Logger,
+	routes rpccore.RoutesMap,
+	eventBus types.EventBusSubscriber) ([]net.Listener, error) {
+
+	config := rpcserver.DefaultConfig()
+	config.MaxBodyBytes = rpcConfig.MaxBodyBytes
+	config.MaxHeaderBytes = rpcConfig.MaxHeaderBytes
+	// If necessary adjust global WriteTimeout to ensure it's greater than
+	// TimeoutBroadcastTxCommit.
+	// See https://github.com/tendermint/tendermint/issues/3435
+	if config.WriteTimeout <= rpcConfig.TimeoutBroadcastTxCommit {
+		config.WriteTimeout = rpcConfig.TimeoutBroadcastTxCommit + 1*time.Second
+	}
+
+	// we may expose the rpc over both a unix and tcp socket
+	listeners, err := listenersFromRPCConfig(rpcConfig)
+	if err != nil {
+		return nil, err
+	}
+	for _, listener := range listeners {
+		mux := http.NewServeMux()
+		registerWebsocketHandler(rpcConfig, mux, routes, logger, eventBus)
+		rpcserver.RegisterRPCFuncs(mux, routes, logger)
+
+		var rootHandler http.Handler = mux
+		if rpcConfig.IsCorsEnabled() {
+			rootHandler = addCORSHandler(rpcConfig, mux)
+		}
+		if rpcConfig.IsTLSEnabled() {
+			go func() {
+				listenerAddr := listener.Addr().String()
```

Comment:
Here and below, would it make sense to do some of this setup before starting the goroutine? That way we can be confident that the log messages will show up before the starter function returns (which can be valuable for understanding the sequence of events when debugging).

Loose principle: Put as little into a goroutine as you can get away with.

Reply:
Good idea, and strongly agree with the principle.

Reply:
@creachadair Is your thought here to move the log statement out of the goroutine as well? I somewhat like having the log statement in the goroutine so that the user can be somewhat sure that the goroutine to start the server actually was scheduled.